### PR TITLE
Implement Data Sources

### PR DESCRIPTION
**Pull request by cbush**
Note to reviewer: disregard lack of E2E for now (i.e. you can't specify which data source to use from the `ingest pages` command, etc)

- Implements the Snooty Data Source ~(TODO: Bypass corp authwall)~
- Snooty AST to MD borrows heavily from https://github.com/mongodben/snooty-ast-to-md
- Implements DevCenter source (must be on VPN or whitelisted IP)
- Implements comparison of incoming pages from source with existing pages in the page store and stores only the created, deleted, or modified
- Briefly switched to ESM for node-fetch, but that interfered with yargs and jest and it just isn't worth it

Right now this is designed to just load the entire data source in memory, then load all of the existing pages, and compare the entire sets. It would probably be safer in the future to rearchitect it to use AsyncGenerators and do everything page-by-page.

### CODE REVIEW

File: ingest/src/DatabaseConnection.test.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import {
   DatabaseConnection,
 } from "./DatabaseConnection";
 import { PageStore, PersistedPage } from "./updatePages";
-import { ChunkStore, EmbeddedChunk } from "./updateChunks";
+import { ChunkStore } from "./updateChunks";
```
Comment (mongodben) [inline ingest/src/DatabaseConnection.test.ts:9]: in the Q&A server we call this content. can we change this to unify on name `content`?

File: ingest/src/DevCenterDataSource.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        const documents = collection.find();
+
+        const pages: Page[] = [];
+        for await (const document of documents) {
+          pages.push({
+            body: document.content,
+            format: "md",
+            sourceName: name,
+            tags: [], // TODO
+            url: document.calculated_slug, // TODO: should these have the baseurl baked in?
```
Comment (mongodben) [inline ingest/src/DevCenterDataSource.ts:40]: > // TODO: should these have the baseurl baked in?

yes it should. consider passing baseUrl as a param to `makeDevCenterDataSource`

File: ingest/src/SnootyDataSource.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  type: "timestamp";
+  data: number;
+};
+
+export type SnootyMetadataEntry = {
+  type: "metadata";
+  data: Record<string, unknown>;
+};
+
+// TODO
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:18]: what's the TODO here?

File: ingest/src/SnootyDataSource.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    The base url used in links etc
+   */
+  baseUrl: string;
+}): Promise<DataSource> => {
+  const fetch =
+    new URL(manifestUrl).protocol === "file:"
+      ? async (url: string) => ({
+          body: fs.createReadStream(new URL(url).pathname),
+        })
+      : (await import("node-fetch")).default;
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:66]: funky! per prev convos, accepting this works even though i don't fully understand what's happening here.

File: ingest/src/SnootyDataSource.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                  const page = await handlePage(
+                    (entry as SnootyPageEntry).data,
+                    { sourceName: name, baseUrl }
+                  );
+                  pages.push(page);
+                })()
+              );
+            case "metadata":
+            case "timestamp":
+              return; // TODO?
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:105]: what might we do here?

File: ingest/src/SnootyDataSource.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                  }`
+                )
+              );
+          }
+        });
+        stream.on("close", () => {
+          resolve();
+        });
+      });
+      await Promise.allSettled(linePromises);
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:120]: interesting use of `Promise.allSettled()`, how you're not returning from it, but using it as a handler for stream. 

no suggestion or question. just a comment 😄

File: ingest/src/SnootyDataSource.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      return pages;
+    },
+  };
+};
+
+const handleAsset = async (asset: SnootyAsset): Promise<void> => {
+  // TODO
+};
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:128]: what might we do here? what's a snooty asset?

File: ingest/src/SnootyDataSource.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
+import fs from "fs";
+import { createInterface } from "readline";
+import { DataSource } from "./DataSource";
+import { Page } from "./updatePages";
+import { snootyAstToMd } from "./snootyAstToMd";
+
+// These types are what's in the snooty manifest jsonl file.
```
Comment (mongodben) [inline ingest/src/SnootyDataSource.ts:7]: for these different types, can you leave brief TSDoc comments explaining what each is?

just based on names, it's not clear to me.

File: ingest/src/commands/pages.ts [hunk 9]
```diff
@@ -1,15 +1,23 @@
...
 // TODO: Option for which data source
 type PagesCommandArgs = Record<string, never>;
 
 const commandModule: CommandModule<PagesCommandArgs> = {
   command: "pages",
   handler: async (args) => {
+    const snootySource = await makeSnootyDataSource({
+      baseUrl: "https://mongodb.com/docs/",
+      manifestUrl:
+        "https://snooty-data-api.docs.prod.mongodb.com/projects/docs/master/documents",
```
Comment (mongodben) [inline ingest/src/commands/pages.ts:14]: when i go to this URL (even on VPN), it redirects me to https://mongodb.com. what's the expected behavior?

File: ingest/src/main.ts [hunk 10]
```diff
@@ -1,4 +1,4 @@
-import * as yargs from "yargs";
+import yargs = require("yargs");
```
Comment (mongodben) [inline ingest/src/main.ts:1]: ??

is this a CSJ thing?

File: ingest/src/snootyAstToMd.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        break;
+      default:
+        break;
+    }
+    return text;
+  }
+
+  // parent nodes
+  switch (node.type) {
+    case "section":
```
Comment (mongodben) [inline ingest/src/snootyAstToMd.ts:31]:
```suggestion
    // recursive block element cases
    case "section":
```

File: ingest/src/snootyAstToMd.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+// Adapted from https://github.com/mongodben/snooty-ast-to-md/blob/main/index.js
+export const snootyAstToMd = (
+  node: SnootyNode,
+  options: SnootyAstToMdOptions,
+  parentHeadingLevel = 0,
+  text = ""
+): string => {
+  if (node.children === undefined) {
```
Comment (mongodben) [inline ingest/src/snootyAstToMd.ts:14]:
```suggestion
  // base cases
  if (node.children === undefined) {
```

File: ingest/src/snootyAstToMd.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        .map((listItem) => snootyAstToMd(listItem, options, parentHeadingLevel))
+        .join("\n");
+      break;
+    case "listItem":
+      text += `- ${node.children
+        .map((child) => snootyAstToMd(child, options, parentHeadingLevel))
+        .join("")}`;
+      break;
+    // TODO: figure out ordered lists
+    // TODO: figure out tables
```
Comment (mongodben) [inline ingest/src/snootyAstToMd.ts:60]: can you make ticket(s) for this?

Comment (mongodben) [reply ingest/src/snootyAstToMd.ts:60]: even if not in MVP, i think it'd be good to have this tracked.

**Review by mongodben — COMMENTED** [review]
looks very good! small comments throughout

File: ingest/src/commands/pages.ts (hunk 9 above)
Comment (cbush) [reply ingest/src/commands/pages.ts:14]: You tell me - I got this URL from you! 😜

Comment (cbush) [reply ingest/src/commands/pages.ts:14]: The other day this was actually loading the entire manifest

File: ingest/src/snootyAstToMd.ts (hunk 11 above)
Comment (cbush) [reply ingest/src/snootyAstToMd.ts:31]: That's what "parent nodes" is doing

File: ingest/src/DatabaseConnection.test.ts (hunk 1 above)
Comment (cbush) [reply ingest/src/DatabaseConnection.test.ts:9]: I think we should discuss that - what makes a "page" NOT "content"? I'm not married to "chunks" but we can probably come up with something descriptive/self-explanatory

File: ingest/src/SnootyDataSource.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+              return linePromises.push(
+                (async () => {
+                  const page = await handlePage(
+                    (entry as SnootyPageEntry).data,
+                    { sourceName: name, baseUrl }
+                  );
+                  pages.push(page);
+                })()
+              );
+            case "metadata":
```
[the remark this answers is not in the document]
Comment (cbush) [reply ingest/src/SnootyDataSource.ts:103]: Fallthrough to next case

File: ingest/src/SnootyDataSource.ts (hunk 7 above)
Comment (cbush) [reply ingest/src/SnootyDataSource.ts:128]: Images etc

**Review by mongodben — CHANGES_REQUESTED** [review]
forgot to request changes on review

File: ingest/src/SnootyDataSource.ts (hunk 4 above)
Comment (cbush) [reply ingest/src/SnootyDataSource.ts:66]: ESM bullshit. However, it turns out I can actually fix how ESMs are loaded in our tsconfig so we don't have to load this dynamically.

File: ingest/src/SnootyDataSource.ts (hunk 5 above)
Comment (cbush) [reply ingest/src/SnootyDataSource.ts:105]: Not sure, all of this has to do with what's in the manifest.

File: ingest/src/main.ts (hunk 10 above)
Comment (cbush) [reply ingest/src/main.ts:1]: Yeah, leftovers from the ESM rabbit hole

File: ingest/src/snootyAstToMd.ts (hunk 13 above)
Comment (cbush) [reply ingest/src/snootyAstToMd.ts:60]: https://jira.mongodb.org/browse/DOCSP-30957

File: ingest/src/snootyAstToMd.ts (hunk 12 above)
Comment (cbush) [reply ingest/src/snootyAstToMd.ts:14]: I have this in "value nodes"

File: ingest/src/DatabaseConnection.test.ts (hunk 1 above)
Comment (mongodben) [reply ingest/src/DatabaseConnection.test.ts:9]: something like 'embedded content' could be more descriptive.

the idea w content and not chunks is if we at some future point also embed other forms of content like imgs or video

Comment (cbush) [reply ingest/src/DatabaseConnection.test.ts:9]: I'm good with EmbeddedContent. Would do in subsequent PR.

Comment (mongodben) [reply ingest/src/DatabaseConnection.test.ts:9]: sgtm. that should also be updated in the Q&A server. can do 1 PR across both. can you make ticket or add to existing one like [DOCSP-30955](https://jira.mongodb.org/browse/DOCSP-30955)

File: ingest/src/commands/pages.ts [hunk 15]
```diff
@@ -1,15 +1,43 @@
 import { CommandModule } from "yargs";
 import { updatePages } from "../updatePages";
+import { makeSnootyDataSource } from "../SnootyDataSource";
+import { makeDevCenterDataSource } from "../DevCenterDataSource";
+import { snootyProjects } from "../snootyProjects";
 
 // TODO: Option for which data source
 type PagesCommandArgs = Record<string, never>;
```
Comment (mongodben) [inline ingest/src/commands/pages.ts:8]: to clarify my understanding: are you not yet implementing passing the data source as args? and right now it's hard coding the data sources in the command itself? 

fine for now, but could you create a ticket for the data sources as args, and note it in a comment?

**Review by mongodben — APPROVED** [review]
lgtm w 1 request

File: ingest/src/commands/pages.ts (hunk 15 above)
Comment (cbush) [reply ingest/src/commands/pages.ts:8]: https://jira.mongodb.org/browse/DOCSP-30988

**Pull request merged**